plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ReusableConnectionFactory.java

```diff
@@ -0,0 +1,184 @@
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorSession;
+import org.gaul.modernizer_maven_annotations.SuppressModernizer;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.time.Duration;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.cache.RemovalCause.EXPLICIT;
+import static io.trino.plugin.jdbc.JdbcErrorCode.JDBC_ERROR;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class ReusableConnectionFactory
+        implements ConnectionFactory, QueryEventListener
+{
+    @GuardedBy("this")
+    private final Cache<String, Connection> connections;
+    private final ConnectionFactory delegate;
+
+    @Inject
+    public ReusableConnectionFactory(@ForReusableConnectionFactory ConnectionFactory delegate)
+    {
+        this(delegate, Duration.ofSeconds(2), 10);
+    }
+
+    ReusableConnectionFactory(ConnectionFactory delegate, Duration duration, long maximumSize)
+    {
+        this.connections = createConnectionsCache(duration, maximumSize);
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    // CacheBuilder.build(CacheLoader) is forbidden, because it does not support eviction for ongoing loads.
+    // In this class, loading is not used and cache is used more as a map. So this is safe.
+    @SuppressModernizer
+    private static Cache<String, Connection> createConnectionsCache(Duration duration, long maximumSize)
+    {
+        requireNonNull(duration, "duration is null");
+        return CacheBuilder.newBuilder()
+                .maximumSize(maximumSize)
+                .expireAfterWrite(duration)
+                .removalListener(ReusableConnectionFactory::onRemoval)
+                .build();
+    }
+
+    private static void onRemoval(RemovalNotification<String, Connection> notification)
+    {
+        if (notification.getCause() == EXPLICIT) {
+            // connection was taken from the cache
+            return;
+        }
+        try {
+            requireNonNull(notification.getValue(), "notification.getValue() is null");
+            notification.getValue().close();
+        }
+        catch (SQLException e) {
+            throw new TrinoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        String queryId = session.getQueryId();
+        Connection connection = getConnection(session, queryId);
+        return new CachedConnection(queryId, connection);
+    }
+
+    private Connection getConnection(ConnectorSession session, String queryId)
+            throws SQLException
+    {
+        Connection connection = connections.asMap().remove(queryId);
+        if (connection != null) {
```

Comment:
what if the session used to create the initial connection is not the same as the one that was used to create the connection?

Reply:
Currently there is no way change the session within a query (other than io.trino.sql.analyzer.StatementAnalyzer.Visitor#analyzeView).  The only thing would be impersonation that could potentially change the session, however we do not support impersonation in Trino for JDBC connectors nor there is no such thing like impersonation during planning.